django/test/testcases.py

```diff
@@ -265,18 +265,25 @@ def _setup_and_call(self, result, debug=False):
                     raise
                 result.addError(self, sys.exc_info())
                 return
-        if debug:
-            super().debug()
-        else:
-            super().__call__(result)
-        if not skipped:
-            try:
-                self._post_teardown()
-            except Exception:
-                if debug:
-                    raise
-                result.addError(self, sys.exc_info())
-                return
+
+        call_teardown = not skipped
+        try:
+            if debug:
+                super().debug()
+            else:
+                super().__call__(result)
+        except BaseException:
```

Comment:
It seems to me that `SkipTest` is the only exception that should be handled when skipping with `debug=True`.

Also, do we need to actually wrap the call `super().__call__(result)`, given that `_Outcome.testPartExecutor` already catches all exceptions, nothing can be raised here (unless it's an exception happening before or after the setup/test/teardown flow).

It's important to note that it's possible to call `SkipError` from setUp and tearDown, so if we want to support the same in `_pre_setup` and `_post_teardown` we should catch that exception here as well...

Reply:
> It seems to me that `SkipTest` is the only exception that should be handled when skipping with `debug=True`.
> 
> Also, do we need to actually wrap the call `super().__call__(result)`, given that `_Outcome.testPartExecutor` already catches all exceptions, nothing can be raised here (unless it's an exception happening before or after the setup/test/teardown flow).

I agree with these comments :point_up: .

> It's important to note that it's possible to call `SkipError` from setUp and tearDown, so if we want to support the same in Django's `_pre_setup` and `_post_teardown` we should catch `SkipTest` exception here as well...

I'm not sure about this. What steps should we take if someone calls `skipTest()` in `_pre_setup()` or `_post_teardown()`?

Reply:
Sorry for not following up earlier.
JFI: pytest actually tries hard getting out of testPartExecutor, which causes `_post_teardown` not being called now with pytest 5.4 (via pytest-django, https://github.com/pytest-dev/pytest-django/issues/824).
IIRC I've done this PR initially when I've looked at Django's code, without having in mind that pytest might end up doing it this way.